src/util-file-decompression.c

```diff
@@ -0,0 +1,207 @@
+ * 02110-1301, USA.
+ */
+
+/** \file
+ *
+ * \author Giuseppe Longo <giuseppe@glongo.it>
+ *
+ * \brief Decompress files transfered via HTTP corresponding to file_data
+ * keyword.
+ *
+ */
+
+#include "suricata-common.h"
+#include "suricata.h"
+
+#include "app-layer-htp.h"
+
+#include "util-file-decompression.h"
+#include "util-file-flash-decompression.h"
+#include "util-misc.h"
+#include "util-print.h"
+
+#define FLASH_ZLIB_MIN_VERSION    0x06
+#define FLASH_LZMA_MIN_VERSION    0x0D
+
+int FileIsFlashFile(const uint8_t *buffer, uint32_t buffer_len)
+{
+    if (buffer_len >= 3 && buffer[1] == 'W' && buffer[2] == 'S') {
+        if (buffer[0] == 'F')
+            return FILE_FLASH_NO_COMPRESSION;
+        else if (buffer[0] == 'C')
+            return FILE_FLASH_ZLIB_COMPRESSION;
+        else if (buffer[0] == 'Z')
+            return FILE_FLASH_LZMA_COMPRESSION;
+        else
+            return FILE_IS_NOT_FLASH;
+    }
+
+    return FILE_IS_NOT_FLASH;
+}
+
+/**
+ * \brief This function decompresses a buffer with zlib/lzma algorithm
+ *
+ * \param buffer compressed buffer
+ * \param buffer_len compressed buffer length
+ * \param decompressed_buffer buffer that store decompressed data
+ * \param decompressed_buffer_len decompressesd data length
+ * \param swf_type decompression algorithm to use
+ * \param decompress_depth how much decompressed data we want to store
+ * \param compress_depth how much compressed data we want to decompress
+ *
+ * \retval 1 if decompression works
+ * \retval 0 an error occured, and event set
+ */
+int FileDecompressFlashFile(const uint8_t *buffer, uint32_t buffer_len,
+                            uint8_t **decompressed_buffer,
+                            uint32_t *decompressed_buffer_len,
+                            DetectEngineThreadCtx *det_ctx,
+                            int swf_type,
+                            uint32_t decompress_depth,
+                            uint32_t compress_depth)
+{
+    int r = 0;
+    uint8_t *compressed_data = NULL;
+    uint8_t *decompressed_data = *decompressed_buffer;
+
+    int compression_type = FileIsFlashFile(buffer, buffer_len);
+    if (compression_type == FILE_FLASH_NO_COMPRESSION) {
+        return 0;
+    }
+
+    uint32_t offset = 0;
+    if (compression_type == FILE_FLASH_ZLIB_COMPRESSION) {
+        /* compressed data start from the 4th bytes */
+        offset = 8;
+    } else if (compression_type == FILE_FLASH_LZMA_COMPRESSION) {
+        /* compressed data start from the 17th bytes */
+        offset = 17;
+    }
+
+    uint32_t compressed_swf_len = 0;
+    if (buffer_len > offset && compress_depth == 0) {
+        compressed_swf_len = buffer_len - offset;
+    } else if (compress_depth > 0 && compress_depth <= buffer_len) {
+        compressed_swf_len = compress_depth;
+    } else if (compress_depth > 0 && compress_depth > buffer_len) {
+        compressed_swf_len = buffer_len;
+    } else {
+        compressed_swf_len = offset;
```

Comment:
not sure I follow how we can get into this condition. Plz add a comment to explain